File: src/synonym_cli/datamuse.py
import asyncio
from os import name as os_name
from rich import print, box
from rich.table import Table
from json import loads
from aiohttp import ClientSession
from urllib.parse import urlencode


class DataMuse:
    def __init__(self, word, max: int = 15):
        self.word = word
        self.max = max
        if len(self.word.split(":")) > 1:
            self.topics = self.word.split(":")[1]
            self.word = self.word.split(":")[0]
        else:
            self.topics = ""

        if os_name == "nt":
            asyncio.set_event_loop_policy(asyncio.WindowsSelectorEventLoopPolicy())
        asyncio.run(self.main())

    async def fetch(self, session, url):
        async with session.get(url) as response:
            if response.status != 200:
                response.raise_for_status()
            return await response.text()

    async def fetch_all(self, session):
        dic = {
            "words": {
                "ml": "🔡[green]similars: ",
                "rel_syn": "🔵[green]synonyms: ",
                "rel_ant": "🟤[green]antonyms: ",
                "rel_trg": "💭[green]evocative: ",
                "sl": "📣[green]sound\[saʊnd]: ",
                "sp": "🧮[green]similar spelling: ",
                "rel_rhy": "👂[green]rhymic: ",
                "rel_jjb": f"[green]___ + [i]{self.word}[/]: ",
                "rel_jja": f"[green][i]{self.word}[/] + ___: ",
            },
            "sug": {"s": f"[green][i]{self.word}[/]...:"},
        }
        tasks = []
        base_url = "https://api.datamuse.com"
        for i in dic:
            for j in dic[i]:
                url_args = urlencode(
                    {j: self.word, "max": self.max, "topics": self.topics}
                )
                url = f"{base_url}/{i}?{url_args}"
                task = asyncio.create_task(self.fetch(session, url))
                tasks.append(task)
        results = await asyncio.gather(*tasks)
        tables = []
        idx = 0
        for i in dic:
            for j in dic[i]:
                word_list = []
                for results[idx] in loads(results[idx]):
                    if results[idx]["word"] != self.word:
                        word_list.append(results[idx]["word"])
                    else:
                        continue
                table = Table(box=box.SIMPLE, expand=True, padding=0)
                table.add_column(dic[i][j])
                row_words = ", ".join(word_list)
                table.add_row("[cyan3]" + row_words)
                if row_words:
                    tables.append(table)
                idx += 1

        topics = self.topics.split(",")
        desc = f'{self.word}{("🪢("+" • ".join(topics)+")") if topics!=[""] else ""}'
        # Iterate through the tables in increments of 3
        for i in range(0, len(tables), 3):
            grid = Table.grid(padding=0, expand=True)
            # Get the current group of 3 tables
            group = tables[i : i + 3]
            # Add the group to a row in the grid
            grid.add_row(*group)
            print(grid)

    async def main(self):
        async with ClientSession() as session:
            await self.fetch_all(session)

File: web/lib/datamuse.ts
export interface DatamuseItem {
  word: string;
  score: number;
  tags?: string[];
}

export interface CategoryResult {
  category: string;
  items: string[];
}

export const CATEGORIES: Record<string, { endpoint: string; param: string }> = {
  Similars: { endpoint: "words", param: "ml" },
  Synonyms: { endpoint: "words", param: "rel_syn" },
  Antonyms: { endpoint: "words", param: "rel_ant" },
  Evocative: { endpoint: "words", param: "rel_trg" },
  "Sounds Like": { endpoint: "words", param: "sl" },
  "Similar Spelling": { endpoint: "words", param: "sp" },
  Rhymes: { endpoint: "words", param: "rel_rhy" },
  "Adjectives for word": { endpoint: "words", param: "rel_jjb" },
  "Nouns for word": { endpoint: "words", param: "rel_jja" },
  Suggestions: { endpoint: "sug", param: "s" },
};

export const CATEGORY_COLORS: Record<string, string> = {
  Similars: "var(--cat-similars)",
  Synonyms: "var(--cat-synonyms)",
  Antonyms: "var(--cat-antonyms)",
  Evocative: "var(--cat-evocative)",
  "Sounds Like": "var(--cat-sounds)",
  "Similar Spelling": "var(--cat-spelling)",
  Rhymes: "var(--cat-rhymes)",
  "Adjectives for word": "var(--cat-adjectives)",
  "Nouns for word": "var(--cat-nouns)",
  Suggestions: "var(--cat-suggestions)",
};

const BASE_URL = "https://api.datamuse.com";

async function fetchCategory(
  word: string,
  category: string,
  topic?: string,
): Promise<CategoryResult> {
  const { endpoint, param } = CATEGORIES[category];
  const queryParams = new URLSearchParams({ [param]: word, max: "20" });
  if (topic) queryParams.append("topics", topic);

  try {
    const res = await fetch(`${BASE_URL}/${endpoint}?${queryParams}`);
    if (!res.ok) throw new Error(res.statusText);
    const data = (await res.json()) as DatamuseItem[];
    return {
      category,
      items: data
        .filter((item) => item.word.toLowerCase() !== word.toLowerCase())
        .map((item) => item.word),
    };
  } catch {
    return { category, items: [] };
  }
}

export async function fetchAll(
  wordInput: string,
  categoryFilter?: string,
): Promise<CategoryResult[]> {
  const [word, topic] = wordInput.split(":");
  if (!word) return [];

  const categoriesToFetch = categoryFilter
    ? [categoryFilter]
    : Object.keys(CATEGORIES);

  const results = await Promise.all(
    categoriesToFetch.map((cat) => fetchCategory(word, cat, topic)),
  );

  return results.filter((r) => r.items.length > 0);
}
